File: clients/macos/SoloAgent/Sources/GameEngine/ExpEngine.swift
import Foundation

/// 经验值引擎 — 被动经验计算 + 专注连击
@MainActor
final class ExpEngine {
    private let playerManager: PlayerManager
    private let bus: GameEventBus

    private(set) var focusStreak: Int = 0
    private var lastStreakBonus: Int = 0
    private(set) var totalPassiveExp: Int = 0

    // Passive exp rules: category -> (base_exp, focus_multiplier_enabled)
    static let passiveExpRules: [String: (baseExp: Int, focusMultiplier: Bool)] = [
        "coding":        (3, true),
        "writing":       (3, true),
        "learning":      (4, true),
        "work":          (2, true),
        "creative":      (3, true),
        "design":        (3, true),
        "reading":       (2, true),
        "browsing":      (1, false),
        "communication": (1, false),
        "social":        (0, false),
        "media":         (0, false),
        "gaming":        (0, false),
        "idle":          (0, false),
        "unknown":       (0, false),
    ]

    // Focus streak bonus thresholds
    static let focusStreakBonuses: [(threshold: Int, bonus: Int)] = [
        (3, 5), (6, 15), (10, 30), (15, 50),
    ]

    init(playerManager: PlayerManager, eventBus: GameEventBus) {
        self.playerManager = playerManager
        self.bus = eventBus

        bus.on(.contextAnalyzed) { [weak self] event in
            self?.onContextAnalyzed(event)
        }
    }

    // MARK: - Context Analyzed Handler

    private func onContextAnalyzed(_ event: GameEvent) {
        let category = event.data["category"] as? String ?? "idle"
        let focusScore = event.data["focus_score"] as? Double ?? 0.0

        guard let rule = Self.passiveExpRules[category] else { return }

        if rule.baseExp <= 0 {
            if focusScore < 0.3 {
                focusStreak = 0
                lastStreakBonus = 0
            }
            return
        }

        // Focus multiplier
        var totalExp = rule.baseExp
        if rule.focusMultiplier && focusScore > 0.5 {
            let multiplier = 1.0 + (focusScore - 0.5) * 2
            totalExp = Int(Double(rule.baseExp) * multiplier)
        }

        // Update streak
        if focusScore >= 0.6 {
            focusStreak += 1
        } else {
            focusStreak = max(0, focusStreak - 1)
        }

        // Streak bonus
        var streakBonus = 0
        for (threshold, bonus) in Self.focusStreakBonuses {
            if focusStreak >= threshold && threshold > lastStreakBonus {
                streakBonus = bonus
                lastStreakBonus = threshold
            }
        }

        if streakBonus > 0 {
            totalExp += streakBonus
            bus.emit(.notificationPush, data: [
                "title": SystemMessages.focusStreakMessage(streak: focusStreak, bonus: streakBonus),
                "style": "exp",
            ])
        }

        // Award exp
        if totalExp > 0 {
            playerManager.gainExp(totalExp, source: "passive:\(category)")
            totalPassiveExp += totalExp
        }
    }
}

File: clients/win/src/lib/game/exp-engine.ts
// ══════════════════════════════════════════════
// Exp Engine — 移植自 macOS ExpEngine.swift
// 被动经验计算 + 专注连击
// ══════════════════════════════════════════════

import type { GameEventBus, GameEvent } from './event-bus'
import type { PlayerManager } from './player-manager'
import { focusStreakMessage } from './system-messages'

// ── Passive Exp Rules ──

interface PassiveExpRule {
  readonly baseExp: number
  readonly focusMultiplier: boolean
}

const PASSIVE_EXP_RULES: Record<string, PassiveExpRule> = {
  coding:        { baseExp: 3, focusMultiplier: true },
  writing:       { baseExp: 3, focusMultiplier: true },
  learning:      { baseExp: 4, focusMultiplier: true },
  work:          { baseExp: 2, focusMultiplier: true },
  creative:      { baseExp: 3, focusMultiplier: true },
  design:        { baseExp: 3, focusMultiplier: true },
  reading:       { baseExp: 2, focusMultiplier: true },
  browsing:      { baseExp: 1, focusMultiplier: false },
  communication: { baseExp: 1, focusMultiplier: false },
  social:        { baseExp: 0, focusMultiplier: false },
  media:         { baseExp: 0, focusMultiplier: false },
  gaming:        { baseExp: 0, focusMultiplier: false },
  idle:          { baseExp: 0, focusMultiplier: false },
  unknown:       { baseExp: 0, focusMultiplier: false },
}

// ── Focus Streak Bonuses ──

const FOCUS_STREAK_BONUSES: readonly { threshold: number; bonus: number }[] = [
  { threshold: 3, bonus: 5 },
  { threshold: 6, bonus: 15 },
  { threshold: 10, bonus: 30 },
  { threshold: 15, bonus: 50 },
]

// ── Engine State ──

export interface ExpEngineState {
  readonly focusStreak: number
  readonly lastStreakBonus: number
  readonly totalPassiveExp: number
}

export function createExpEngineState(): ExpEngineState {
  return { focusStreak: 0, lastStreakBonus: 0, totalPassiveExp: 0 }
}

// ── Core Logic ──

export interface ExpEngineResult {
  readonly state: ExpEngineState
  readonly expGained: number
  readonly streakBonus: number
  readonly streakMessage: string | null
}

export function processContextAnalyzed(
  state: ExpEngineState,
  category: string,
  focusScore: number,
): ExpEngineResult {
  const rule = PASSIVE_EXP_RULES[category]

  if (!rule || rule.baseExp <= 0) {
    // Non-productive: decay streak if low focus
    if (focusScore < 0.3) {
      return {
        state: { ...state, focusStreak: 0, lastStreakBonus: 0 },
        expGained: 0,
        streakBonus: 0,
        streakMessage: null,
      }
    }
    return { state, expGained: 0, streakBonus: 0, streakMessage: null }
  }

  // Focus multiplier
  let totalExp = rule.baseExp
  if (rule.focusMultiplier && focusScore > 0.5) {
    const multiplier = 1.0 + (focusScore - 0.5) * 2
    totalExp = Math.floor(rule.baseExp * multiplier)
  }

  // Update streak
  let focusStreak: number
  if (focusScore >= 0.6) {
    focusStreak = state.focusStreak + 1
  } else {
    focusStreak = Math.max(0, state.focusStreak - 1)
  }

  // Streak bonus
  let streakBonus = 0
  let lastStreakBonus = state.lastStreakBonus
  for (const { threshold, bonus } of FOCUS_STREAK_BONUSES) {
    if (focusStreak >= threshold && threshold > lastStreakBonus) {
      streakBonus = bonus
      lastStreakBonus = threshold
    }
  }

  let streakMessage: string | null = null
  if (streakBonus > 0) {
    totalExp += streakBonus
    streakMessage = focusStreakMessage(focusStreak, streakBonus)
  }

  return {
    state: {
      focusStreak,
      lastStreakBonus,
      totalPassiveExp: state.totalPassiveExp + totalExp,
    },
    expGained: totalExp,
    streakBonus,
    streakMessage,
  }
}

// ── Bus Integration ──

export function setupExpEngine(
  bus: GameEventBus,
  playerManager: PlayerManager,
  getState: () => ExpEngineState,
  setState: (s: ExpEngineState) => void,
): string {
  return bus.on('contextAnalyzed', (event: GameEvent) => {
    const category = (event.data.category as string) ?? 'idle'
    const focusScore = (event.data.focus_score as number) ?? 0.0

    const result = processContextAnalyzed(getState(), category, focusScore)
    setState(result.state)

    if (result.streakMessage) {
      bus.emit('notificationPush', {
        title: result.streakMessage,
        style: 'exp',
      })
    }

    if (result.expGained > 0) {
      playerManager.gainExp(result.expGained, `passive:${category}`)
    }
  })
}
